api/oc_ri.c

```diff
@@ -1562,68 +1422,423 @@ oc_ri_invoke_coap_entity_handler(coap_make_response_ctx_t *ctx,
+
+    /* Found the matching resource object. Now verify that:
+     * 1) the selected interface is one that is supported by
+     *    the resource, and,
+     * 2) the selected interface supports the request method.
+     *
+     * If not, return a 4.00 response.
+     */
+    if (((iface_mask & ~cur_resource->interfaces) != 0) ||
+        !oc_ri_interface_supports_method(iface_mask, method)) {
+      bitmask_code |= (BITMASK_CODE_BAD_REQUEST | BITMASK_CODE_FORBIDDEN);
 #ifdef OC_SECURITY
-  else if (not_authorized) {
-    OC_WRN("ocri: Subject not authorized");
-    /* If the requestor (subject) does not have access granted via an
-     * access control entry in the ACL, then it is not authorized to
-     * access the resource. A 4.01 response is sent.
+      oc_audit_log(endpoint->device, "COMM-1", "Operation not supported", 0x40,
+                   2, NULL, 0);
+#endif /* OC_SECURITY */
+    }
+  } else {
+    bitmask_code |= BITMASK_CODE_NOT_FOUND;
+  }
+
+  if (!bitmask_code && !oc_rep_decoder_is_supported_content_format(
+                         ctx->preparsed_request_obj->cf)) {
+    bitmask_code |= BITMASK_CODE_BAD_REQUEST;
+  }
+
+  if (!bitmask_code && cur_resource &&
+      (!get_resource_is_collection(ctx->preparsed_request_obj)) &&
+      !oc_resource_get_method_handler(cur_resource, method, NULL)) {
+    bitmask_code |= BITMASK_CODE_METHOD_NOT_ALLOWED;
+  }
+
+#ifdef OC_SECURITY
+  if (cur_resource && !oc_sec_check_acl(method, cur_resource, endpoint)) {
+    /* If resource is a coaps:// resource, then query ACL to check if
+     * the requestor (the subject) is authorized to issue this request to
+     * the resource.
      */
-    response_buffer.response_length = 0;
-    response_buffer.code = oc_status_code_unsafe(OC_STATUS_UNAUTHORIZED);
+    oc_ri_audit_log(method, cur_resource, endpoint);
+    bitmask_code |= BITMASK_CODE_UNAUTHORIZED;
   }
 #endif /* OC_SECURITY */
-  else {
-    if (not_modified) {
-      response_buffer.response_length = 0;
-      response_buffer.code = oc_status_code_unsafe(OC_STATUS_NOT_MODIFIED);
+
+  return ri_validate_coap_set_response(ctx->response, endpoint, bitmask_code);
+}
+
+#ifdef OC_HAS_FEATURE_ETAG
+typedef struct
+{
+  const coap_packet_t *request;
+  const oc_ri_preparsed_request_obj_t *preparsed_request_obj;
+  bitmask_code_t bitmask_code;
+  const oc_endpoint_t *endpoint;
+  oc_method_t method;
+  oc_interface_mask_t iface_mask;
+} ri_invoke_coap_set_etag_in_t;
+
+static bitmask_code_t
+ri_invoke_coap_set_etag(const ri_invoke_coap_set_etag_in_t *in,
+                        oc_request_t *request_obj)
+{
+  bitmask_code_t bitmask_code = in->bitmask_code;
+  if (bitmask_code || in->preparsed_request_obj->cur_resource == NULL ||
+      in->method != OC_GET ||
+      !(get_resource_is_collection(in->preparsed_request_obj) ||
+        in->preparsed_request_obj->cur_resource->get_handler.cb != NULL)) {
+    return bitmask_code;
```

Comment:
the retuned value of `ri_invoke_coap_set_etag` is OR-ed with the current bitmask_code so we don't need to return the original value, we can return 0 and simplify the code (remove the bitmask_code member from the `ri_invoke_coap_set_etag` struct).